latch/registry/table.py

```diff
@@ -0,0 +1,257 @@
+                }
+            }
+            """,
+            variables={"argTableId": self.id},
+        )["catalogExperiment"]["catalogExperimentColumnDefinitionsByExperimentId"][
+            "nodes"
+        ]
+
+        return self._columns
+
+    @property
+    def row_type(self):
+        if self._row_type is not None:
+            return self._row_type
+
+        fields = []
+        for column in self.columns:
+            cleaned = clean(column["key"])
+            column_db_type = column["type"]
+
+            fields.append(
+                (
+                    cleaned,
+                    to_python_type(
+                        column_db_type["type"],
+                        cleaned,
+                        column_db_type["allowEmpty"],
+                    ),
+                )
+            )
+
+        return make_dataclass(
+            clean(f"{self.name}_type"),
+            fields,
+            bases=(Row,),
+            frozen=True,
+        )
+
+    def list(self) -> List[Row]:
+        rows = gql.execute(
+            """
+            query listRowsQuery($argTableId: BigInt!) {
+                catalogExperiment(id: $argTableId) {
+                    catalogSamplesByExperimentId(condition: { removed: false }) {
+                        nodes {
+                            id
+                            name
+                            catalogSampleColumnDataBySampleId {
+                                nodes {
+                                    data
+                                    key
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            """,
+            {"argTableId": self.id},
+        )["catalogExperiment"]["catalogSamplesByExperimentId"]["nodes"]
+
+        output = []
+
+        for row in rows:
+            row_data = row["catalogSampleColumnDataBySampleId"]["nodes"]
+            values = {"id": row["id"], "name": row["name"]}
+
+            try:
+                for column in self.columns:
+                    key = clean(column["key"])
+                    typ = column["type"]
+
+                    for data_point in row_data:
+                        if clean(data_point["key"]) != key:
+                            continue
+                        values[key] = to_python_literal(
+                            data_point["data"],
+                            typ["type"],
+                        )
+                        break
+
+                    if key not in values:
+                        if not typ["allowEmpty"]:
+                            raise ValueError(
+                                f"no value provided for required column {key} -"
+                                " skipping"
+                            )
+                        values[key] = None
+            except ValueError as e:
+                print(f"row {row['id']} ({row['name']}) is invalid - skipping")
```

Comment:
collect errors and return them or just raise here

print is never how you do error handling